### PR TITLE
[Fix #11191] Add new `Lint/NumericOperationWithConstantResult` cop

### LINKED ISSUE #11191: New cop that detects operations with constant results
Inspired by https://hexdocs.pm/credo/Credo.Check.Warning.OperationWithConstantResult.html

It should check for the code like
```ruby
x * 1
x * 0
x / 1
x ** 0
x ** 1
x % 1
```

This is almost always a mistake (or just left after debugging, probably) or just redundant code.

### PR DESCRIPTION
**Pull request by Zopolis4**
Fixes #11191.

Modifications were made to `Lint/BinaryOperatorWithIdenticalOperands` in order to stop it from acting on arithmetic operations, as to avoid doubling up. 

At any rate, `Lint/BinaryOperatorWithIdenticalOperands` would benefit from a rework at some point to support autocorrect, so limit the extent of that task anyways.

-----------------

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/lint/numeric_operation_with_constant_result.rb [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      #   # bad
+      #   x /= x
+      #   x **= 0
+      #
+      #   # good
+      #   x = 1
+      #
+      class NumericOperationWithConstantResult < Base
+        extend AutoCorrector
+        MSG = 'Do not apply numeric operations with constant results to variables.'
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/numeric_operation_with_constant_result.rb:49]: I think "Numeric operation with a constant result detected." would be a slightly better message here.

File: lib/rubocop/cop/lint/numeric_operation_with_constant_result.rb [hunk 2]
```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # Certain numeric operations have a constant result, usually 0 or 1.
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/numeric_operation_with_constant_result.rb:6]: You should probably mention the other cop that's a relative to this one here, so people understand well why the scope is so narrow.

**Comment by bbatsov** [comment]
Looks good to me overall, but I left a couple of small remarks.

**Comment by Zopolis4** [comment]
Thanks for the review! I've addressed both of your comments.

File: lib/rubocop/cop/lint/numeric_operation_with_constant_result.rb [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+      # Subtracting a number from itself or multiplying it by 0 will always return 0.
+      # Additionally, a variable modulo 0 or itself will always return 0.
+      # Dividing a number by itself or raising it to the power of 0 will always return 1.
+      # As such, they can be replaced with that result.
+      # These are probably leftover from debugging, or are mistakes.
+      # Other numeric operations that are similarly leftover from debugging or mistakes
+      # are handled by Lint/UselessNumericOperation.
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/numeric_operation_with_constant_result.rb:13]: There's also `UselessNumericOperation`. Probably we should update its docs as well, as it's kind of similar to this one. Too many cops! 😆

**Pull request merged**